components/script/dom/csp.rs

```diff
@@ -0,0 +1,305 @@
+                // Step 3.1: If target is not null, and global is a Window,
+                // and target’s shadow-including root is not global’s associated Document, set target to null.
+                if let Some(window) = self.downcast::<Window>() {
+                    // If a node is connected, its owner document is always the shadow-including root.
+                    // If it isn't connected, then it also doesn't have a corresponding document, hence
+                    // it can't be this document.
+                    if event_target.upcast::<Node>().owner_document() != window.Document() {
+                        return None;
+                    }
+                }
+                Some(event_target)
+            });
+            let target = match target {
+                // Step 3.2: If target is null:
+                None => {
+                    // Step 3.2.2: If target is a Window, set target to target’s associated Document.
+                    if let Some(window) = self.downcast::<Window>() {
+                        Trusted::new(window.Document().upcast())
+                    } else {
+                        // Step 3.2.1: Set target to violation’s global object.
+                        Trusted::new(self.upcast())
+                    }
+                },
+                Some(event_target) => Trusted::new(event_target.upcast()),
+            };
+            // Step 3: Queue a task to run the following steps:
+            let task =
+                CSPViolationReportTask::new(Trusted::new(self), target, report, violation.policy);
+            self.task_manager()
+                .dom_manipulation_task_source()
+                .queue(task);
+        }
+    }
+}
+
+/// <https://www.w3.org/TR/CSP/#initialize-document-csp>
+pub(crate) fn parse_csp_list_from_metadata(headers: &Option<Serde<HeaderMap>>) -> Option<CspList> {
+    // TODO: Implement step 1 (local scheme special case)
+    let headers = headers.as_ref()?;
+    let mut csp = headers.get_all("content-security-policy").iter();
+    // This silently ignores the CSP if it contains invalid Unicode.
+    // We should probably report an error somewhere.
+    let c = csp.next().and_then(|c| c.to_str().ok())?;
+    let mut csp_list = CspList::parse(c, PolicySource::Header, PolicyDisposition::Enforce);
+    for c in csp {
+        let c = c.to_str().ok()?;
+        csp_list.append(CspList::parse(
+            c,
+            PolicySource::Header,
+            PolicyDisposition::Enforce,
+        ));
+    }
+    let csp_report = headers
+        .get_all("content-security-policy-report-only")
+        .iter();
+    // This silently ignores the CSP if it contains invalid Unicode.
+    // We should probably report an error somewhere.
+    for c in csp_report {
+        let c = c.to_str().ok()?;
+        csp_list.append(CspList::parse(
+            c,
+            PolicySource::Header,
+            PolicyDisposition::Report,
+        ));
+    }
+    Some(csp_list)
```

Comment:
I think it's a bit confusing to have both `csp` and `c` variables here, so please use names that are more descriptive. Perhaps "csp_header" and "csp" or "policy" work better here.

Reply:
I copied the method as-is. When we refactor this code to handle headers properly, we can pick this up? I rather avoid method rewriting when copying, if that's possible.